jetstream_pt/ray_engine_worker.py

```diff
@@ -0,0 +1,851 @@
+    self.cache_sharding = self.y_sharding
+
+    self._compiled_call_model_prefill = jax.jit(
+        self._call_model_prefill,
+        donate_argnums=(1, 2),
+        out_shardings=(self.replicated, self.cache_sharding),
+    )
+    self._compiled_insert = jax.jit(
+        self._insert,
+        donate_argnums=(0, 1),
+        out_shardings=(
+            self.replicated,
+            self.cache_sharding,
+            self.replicated,
+            self.replicated,
+            self.replicated,
+            self.replicated,
+            self.replicated,
+        ),
+    )
+
+    self._compiled_call_model_generate = jax.jit(
+        self._call_model_generate,
+        donate_argnums=(2, 3, 4, 5, 6, 7),
+        out_shardings=(
+            self.replicated,
+            self.cache_sharding,
+            self.replicated,
+            self.replicated,
+            self.replicated,
+            self.replicated,
+            self.replicated,
+        ),
+    )
+    self._lock = threading.RLock()
+
+  # pylint: disable-next=all
+  def sharding_by_name(self, name):
+
+    # This allows easier way to edit shardings
+    """
+    for key, val in self.env._data.experimental_sharding_axis_override.items():
+      if name.endswith(key):
+        return self.env.sharding_by_axis(val)
+    """
+
+    if "weight_scaler" in name:
+      return self.x_sharding
+    if "tok_embeddings." in name:
+      return self.y_sharding
+    if "attention." in name:
+      if "wo" in name:
+        return self.y_sharding
+      return self.x_sharding
+    if "feed_forward." in name:
+      if "w2" in name:
+        return self.y_sharding
+      return self.x_sharding
+    if "output" in name:
+      return self.x_sharding
+    return self.replicated
+
+  # pylint: disable-next=all
+  def init_decode_state(
+      self,
+  ) -> DecodeState:
+    caches_obj = self.env.make_caches_generate()
+    caches = [c.state() for c in caches_obj]
+    scalers = []
+    if self.env.enable_kv_quantization:
+      scalers = [c.scalers() for c in caches_obj]
+    return DecodeState(
+        jnp.zeros((self.env.batch_size, 1), dtype=jnp.int32),
+        caches,
+        scalers,
+        self.env.max_input_sequence_length,
+        jnp.zeros((self.env.batch_size, 1), dtype=jnp.int32),
+        jnp.zeros((self.env.batch_size,), dtype=jnp.int32),  # input pos
+        jnp.full(
+            (self.env.batch_size, self.env.cache_sequence_length),
+            float("-inf"),
+            dtype=self.default_dtype,
+        ),
+    )
+
+  def print_mem_usage(self):
+    """Print current mem usage"""
+    fmt_size = functools.partial(humanize.naturalsize, binary=True)
+
+    for d in jax.local_devices():
+      stats = d.memory_stats()
+      used = stats["bytes_in_use"]
+      limit = stats["bytes_limit"]
+      print(
+          f"memory using {fmt_size(used)} / {fmt_size(limit)} ({used/limit:%}) on {d}"
+      )
+
+  def init_decode_state_ray(
```

Comment:
nit: maybe this can be `init_decode_state` and the `init_decode_state` from above can become `_init_decode_state`?

Reply:
Ah okay that makes a lot of sense!